internal/dynamic/plugin/subprocess/subprocess.go

```diff
@@ -0,0 +1,211 @@
+		option(s)
+	}
+	return s, nil
+}
+
+func (s *subprocess) Start() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.cmd != nil {
+		return ErrProcessAlreadyStarted
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	cmd := exec.CommandContext(ctx, s.cmdArgs[0], s.cmdArgs[1:]...)
+	if s.env != nil {
+		cmd.Env = os.Environ() // Inherit current environment
+		for k, v := range s.env {
+			cmd.Env = append(s.cmd.Env, fmt.Sprintf("%s=%s", k, v))
+		}
+	}
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		cancel()
+		return fmt.Errorf("failed to create stdout pipe: %w", err)
+	}
+	stderrPipe, err := cmd.StderrPipe()
+	if err != nil {
+		stdoutPipe.Close()
+		cancel()
+		return fmt.Errorf("failed to create stderr pipe: %w", err)
+	}
+	if err := cmd.Start(); err != nil {
+		stdoutPipe.Close()
+		stderrPipe.Close()
+		cancel()
+		return fmt.Errorf("failed to start command: %w", err)
+	}
+	s.logger.Debugf("Subprocess started with PID: %d", cmd.Process.Pid)
+	s.wg.Add(3) // For stdout, stderr, and process wait goroutines
+	go s.readOutput(stdoutPipe, false)
+	go s.readOutput(stderrPipe, true)
+	go func() {
+		defer s.wg.Done()
+		err := cmd.Wait()
+		s.logger.Debugf("Subprocess with PID %d exited with error: %v", cmd.Process.Pid, err)
+	}()
+	s.cmd = cmd
+	s.cancel = cancel
+	return nil
+}
+
+func (s *subprocess) readOutput(pipe io.Reader, isStderr bool) {
+	defer s.wg.Done()
+	src := map[bool]string{false: "stdout", true: "stderr"}[isStderr]
+	log := s.logger.With("source", src)
+	scanner := bufio.NewScanner(pipe)
+	hook := func(line string) {}
+	if !isStderr && s.stdoutHook != nil {
+		hook = s.stdoutHook
+	}
+	if isStderr && s.stderrHook != nil {
+		hook = s.stderrHook
+	}
+	for scanner.Scan() {
+		line := scanner.Text()
+		hook(line)
+		log.Infof("%s", line)
+	}
+	if err := scanner.Err(); err != nil && err != io.EOF {
+		log.Warnf("error reading from subprocess: %v", err)
+	}
+}
+
+func (s *subprocess) IsRunning() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.cmd == nil {
+		return false
+	}
+	if s.cmd.ProcessState != nil && s.cmd.ProcessState.Exited() {
+		return false
+	}
+	return true
+}
+
+func (s *subprocess) Close(ctx context.Context) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.cmd == nil || s.cancel == nil {
+		s.logger.Tracef("Close called on a subprocess that is not running or already closed.")
+		return nil // Not running or already closed
+	}
+
+	s.logger.Debugf("Attempting to gracefully shut down subprocess with PID %d...", s.cmd.Process.Pid)
+	if s.cmd.Process != nil {
+		if err := s.cmd.Process.Signal(os.Interrupt); err != nil {
+			s.logger.Tracef("Failed to send interrupt signal to subprocess PID %d: %v. Attempting to kill.", s.cmd.Process.Pid, err)
+			if err := s.cmd.Process.Kill(); err != nil {
+				s.logger.Warnf("Failed to kill subprocess PID %d: %v", s.cmd.Process.Pid, err)
+			}
+		}
+	}
+	// Use the provided context for waiting for the process to exit
+	waitChan := make(chan struct{})
```

Comment:
I think the most idiotic (iconic) variable name in this case is done :)


```
michal@mac go1.24.2$ rg -F 'done := make(' | wc -l
     221
michal@mac go1.24.2$ rg -F 'wait := make(' | wc -l
       2
michal@mac go1.24.2$ rg -F 'waitCh := make(' | wc -l
       1
michal@mac go1.24.2$ rg -F 'doneCh := make(' | wc -l
       1
michal@mac go1.24.2$ rg -F 'waitChan := make(' | wc -l
       0
```

Reply:
Ha nice, fixed :)